### PR TITLE
Deprecate Sonata Cache in favor of PSR Cache

### LINKED ISSUE #627: Use PSR Cache
## Feature Request

We should remove the Sonata Cache dependency and replace it with the more common PSR Cache.

### PR DESCRIPTION
**Pull request by core23**
## Subject

I am targeting this branch, because this is BC.

Closes #627 

## Changelog

```markdown
### Added
- Added support for PSR Cache

### Deprecated
- Deprecate Sonata Cache in favor of PSR Cache
```

### CODE REVIEW

File: tests/Templating/Helper/BlockHelperTest.php [hunk 1]
```diff
@@ -14,21 +14,75 @@
...
+            ->willReturn($service);
+
+        $blockRenderer = $this->createMock(BlockRendererInterface::class);
+
+        $block = $this->createMock(BlockInterface::class);
+
+        $blockContext = $this->createMock(BlockContextInterface::class);
+        $blockContext->method('getBlock')
+            ->willReturn($block);
+        $blockContext->method('getSetting')->willReturnCallback(static function ($key) {
```
Comment (phansys) [inline tests/Templating/Helper/BlockHelperTest.php]:
```suggestion
        $blockContext->method('getSetting')->willReturnCallback(static function (string $key) {
```

Comment (phansys) [reply tests/Templating/Helper/BlockHelperTest.php]: IMO, we enforce type declarations here too. At least, where is possible.

File: src/Templating/Helper/BlockHelper.php [hunk 2]
```diff
@@ -84,15 +91,41 @@ class BlockHelper extends Helper
...
-    public function __construct(BlockServiceManagerInterface $blockServiceManager, array $cacheBlocks, BlockRendererInterface $blockRenderer,
-                                BlockContextManagerInterface $blockContextManager, EventDispatcherInterface $eventDispatcher,
-                                CacheManagerInterface $cacheManager = null, HttpCacheHandlerInterface $cacheHandler = null, Stopwatch $stopwatch = null)
-    {
+    /**
+     * @param CacheManagerInterface|CacheItemPoolInterface|null $cacheManagerOrCachePool
+     */
+    public function __construct(
+        BlockServiceManagerInterface $blockServiceManager,
+        array $cacheBlocks,
```
Comment (phansys) [inline src/Templating/Helper/BlockHelper.php]: We should document the types allowed in this array, if possible.

File: src/Templating/Helper/BlockHelper.php [hunk 3]
```diff
@@ -84,15 +91,41 @@ class BlockHelper extends Helper
...
         $this->blockRenderer = $blockRenderer;
         $this->eventDispatcher = $eventDispatcher;
-        $this->cacheManager = $cacheManager;
+
+        if ($cacheManagerOrCachePool instanceof CacheManagerInterface) {
+            @trigger_error(
+                sprintf(
+                    'Passing %s as argument 6 to %s::%s() is deprecated since sonata-project/block-bundle 3.x and will throw a \TypeError as of 4.0. You must pass an instance of %s instead.',
+                    CacheManagerInterface::class,
+                    self::class,
```
Comment (phansys) [inline src/Templating/Helper/BlockHelper.php]:
```suggestion
                    static::class,
```

File: src/Templating/Helper/BlockHelper.php [hunk 4]
```diff
@@ -488,4 +497,83 @@ protected function getCacheService(BlockInterface $block, array &$stats = null)
 
         return $this->cacheManager->getCacheService($cacheServiceId);
     }
+
+    private function getCachedBlock(BlockContextInterface $blockContext, BlockServiceInterface $service, array &$stats): ?Response
```
Comment (phansys) [inline src/Templating/Helper/BlockHelper.php]: Could we document `$stats` argument?

File: tests/Templating/Helper/BlockHelperTest.php [hunk 5]
```diff
@@ -14,21 +14,78 @@
...
```
Comment (phansys) [inline tests/Templating/Helper/BlockHelperTest.php]:
```suggestion
        $blockContext->method('getSetting')->willReturnCallback(static function (string $key) {
```

File: tests/Templating/Helper/BlockHelperTest.php [hunk 6]
```diff
@@ -14,21 +14,78 @@
...
 use Sonata\BlockBundle\Templating\Helper\BlockHelper;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 use Symfony\Component\HttpFoundation\Response;
 
 final class BlockHelperTest extends TestCase
 {
+    /**
+     * @group legacy
+     */
+    public function testRenderWithCachedBlock()
```
Comment (phansys) [inline tests/Templating/Helper/BlockHelperTest.php]:
```suggestion
    public function testRenderWithCachedBlock(): void
```

File: src/Templating/Helper/BlockHelper.php (hunk 2 above)
Comment (core23) [reply src/Templating/Helper/BlockHelper.php]: Not related to this PR, but I can fix this

Comment (core23) [reply src/Templating/Helper/BlockHelper.php]: Fixed

File: src/Templating/Helper/BlockHelper.php (hunk 4 above)
Comment (core23) [reply src/Templating/Helper/BlockHelper.php]: Fixed

**Pull request merged**